src/Lecture2.hs

```diff
@@ -158,16 +178,85 @@ You're free to define any helper functions.
 -}
 
 -- some help in the beginning ;)
+data Chest = Chest
+    { chestGold     :: Int
+    , chestTreasure :: Int
+    }
+    deriving (Show)
+
+add :: Chest -> Chest -> Chest
+add c1 c2 = Chest (chestGold c1 + chestGold c2)
+                  (chestTreasure c1 + chestTreasure c2)
+
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
+    { knightHealth     :: Int
+    , knightAttack     :: Int
+    , knightEndurance  :: Int
+    , knightExperience :: Int
+    , knightChest      :: Chest
     }
+    deriving (Show)
 
-dragonFight = error "TODO"
+data Colour
+    = Red
+    | Black
+    | Green
+    deriving (Show, Eq)
+
+data Dragon = Dragon
+    { dragonColour    :: Colour
+    , dragonChest     :: Chest
+    , dragonHealth    :: Int
+    , dragonAttack    :: Int
+    , dragonEndurance :: Int
+    }
+    deriving (Show)
+
+dragonExperience :: Dragon -> Int
+dragonExperience dragon =
+    case (dragonColour dragon) of
+        Red   -> 100
+        Black -> 150
+        Green -> 250
+
+data FightOutcome
+    = KnightWins
+    | KnightDies
+    | KnightRunsAway
+    deriving (Show)
+
+dragonFight :: Knight -> Dragon -> FightOutcome
+dragonFight knight dragon = go 1 knight dragon
+    where go strike knight' dragon'
+            | knightEndurance knight' <= 0 = KnightRunsAway
+            | dragonHealth dragon' <= 0    = KnightWins
+            | knightHealth knight' <= 0    = KnightDies
+            | otherwise                    = go (strike + 1) knight'' dragon''
+              where (knight'', dragon'') = fightRound strike knight' dragon'
+
+fightRound :: Int -> Knight -> Dragon -> (Knight, Dragon)
+fightRound strike k d = (knight', dragon')
+    where knight'
+            | isMultipleOf10 strike = k { knightHealth    = knightHealth k - dragonAttack d
+                                        , knightEndurance = knightEndurance k - 1}
+            | otherwise             = k { knightEndurance = knightEndurance k - 1}
+          dragon'                   = d { dragonHealth = dragonHealth d - knightAttack k}
+          isMultipleOf10 s = s `mod` 10 == 0
+
+updateFighters :: Knight -> Dragon -> FightOutcome -> (Maybe Knight, Maybe Dragon)
```

Comment:
This function is not used anywhere... but I thought we kind of needed it?
How else would we account for "updating" experience and treasures in case of a knight's victory?